kartothek/serialization/_parquet.py

```diff
@@ -175,6 +189,57 @@ def restore_dataframe(
         else:
             return df
 
+    @classmethod
+    def restore_dataframe(
+        cls,
+        store: KeyValueStore,
+        key: str,
+        filter_query: Optional[str] = None,
+        columns: Optional[Iterable[str]] = None,
+        predicate_pushdown_to_io: bool = True,
+        categories: Optional[Iterable[str]] = None,
+        predicates: Optional[PredicatesType] = None,
+        date_as_object: bool = False,
+    ) -> pd.DataFrame:
+        # XXX: We have been seeing weired `IOError` while reading Parquet files from Azure Blob Store,
+        # thus, we implement retries at this point. This code should not live forever, it should be removed
+        # once the underlying cause has been resolved
+        for nb_retry in range(MAX_NB_RETRIES):
```

Comment:
Is there a reason you implemented the retry logic itself? I would suggest to use an established library like [tenacity](https://github.com/jd/tenacity). This also has nice features such as being able to overwrite the wait function in tests.

Also I think you should add tests for the retry logic.

Reply:
I am very strongly against exposing this as some kind of configuration. Current investigations hint towards a bug in our azure-storage interface and/or the iobuffer interface.

Considering our simple requirement, I am in favour of this approach to not introduce tenacity as a core dependency

Reply:
Fair enough. But especially with the self baked retries I think tests are important because the retry now contains untested code logic.

Reply:
Fair point about the testing. I would suggest to test this by patching the `_restore_dataframe` method to raise once but succeeds on the second try. While this test would be very close to the implementation, I believe this is justified for this case

cc @NeroCorleone

Reply:
We expect the following behavior:
* Retry on `OSError` `MAX_NB_RETRIES` times or less if success. Log on retry
* Don't retry on other exceptions
* Raise `ParquetReadError` after `MAX_NB_RETRIES`
* Retry times use exponential backoff

I think all of these could be individual tests.